pkg/solana/chainreader/chain_reader.go

```diff
@@ -495,3 +512,141 @@ func applyIndexedFieldTuple(lookup map[string]uint64, subKeys [4][]string, conf
+	returnSliceVal := reflect.ValueOf(returnVal)
+	if returnSliceVal.Kind() == reflect.Ptr {
+		returnSliceVal = returnSliceVal.Elem()
+		if returnSliceVal.Kind() == reflect.Ptr {
+			returnSliceVal = returnSliceVal.Elem()
+		}
+	}
+	if returnSliceVal.Kind() != reflect.Slice {
+		return fmt.Errorf(
+			"for contract %q read %q: expected `returnVal` to be a slice, got %s",
+			values.contract, values.multiRead[0], returnSliceVal.Kind(),
+		)
+	}
+
+	elemType := returnSliceVal.Type().Elem()
+	for _, d := range data {
+		var wrapper fee_quoter.BillingTokenConfigWrapper
+		if err = wrapper.UnmarshalWithDecoder(bin.NewBorshDecoder(d)); err != nil {
+			return fmt.Errorf(
+				"for contract %q read %q: failed to unmarshal account data: %w",
+				values.contract, values.multiRead[0], err,
+			)
+		}
+
+		newElem := reflect.New(elemType).Elem()
+
+		valueField := newElem.FieldByName("Value")
+		if !valueField.IsValid() {
+			return fmt.Errorf(
+				"for contract %q read %q: struct type missing `Value` field",
+				values.contract, values.multiRead[0],
+			)
+		}
+		valueField.Set(reflect.ValueOf(big.NewInt(0).SetBytes(wrapper.Config.UsdPerToken.Value[:])))
+
+		timestampField := newElem.FieldByName("Timestamp")
+		if !timestampField.IsValid() {
+			return fmt.Errorf(
+				"for contract %q read %q: struct type missing `Timestamp` field",
+				values.contract, values.multiRead[0],
+			)
+		}
+
+		// nolint:gosec
+		// G115: integer overflow conversion int64 -&gt; uint32
+		timestampField.Set(reflect.ValueOf(uint32(wrapper.Config.UsdPerToken.Timestamp)))
+
+		returnSliceVal.Set(reflect.Append(returnSliceVal, newElem))
+	}
+
+	return nil
+}
+
+func (s *ContractReaderService) getPDAsForGetTokenPrices(params any, values readValues) ([]solana.PublicKey, error) {
+	val := reflect.ValueOf(params)
+	if val.Kind() == reflect.Ptr {
+		val = val.Elem()
+	}
+	if val.Kind() != reflect.Struct {
+		return nil, fmt.Errorf(
+			"for contract %q read %q: expected `params` to be a struct, got %s",
+			values.contract, values.multiRead[0], val.Kind(),
+		)
+	}
+
+	field := val.FieldByName("Tokens")
+	if !field.IsValid() {
+		return nil, fmt.Errorf(
+			"for contract %q read %q: no field named 'Tokens' found in params",
+			values.contract, values.multiRead[0],
+		)
+	}
+
+	tokens, ok := field.Interface().(*[][32]uint8)
+	if !ok {
+		return nil, fmt.Errorf(
+			"for contract %q read %q: 'Tokens' field is not of type *[][32]uint8",
+			values.contract, values.multiRead[0],
+		)
+	}
+
+	programAddress, err := solana.PublicKeyFromBase58(values.address)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"for contract %q read %q: %w (could not parse program address %q)",
+			values.contract, values.multiRead[0], types.ErrInvalidConfig, values.address,
+		)
+	}
+
+	// Build the PDA addresses for all tokens.
+	var pdaAddresses []solana.PublicKey
+	for _, token := range *tokens {
+		tokenAddr := solana.PublicKeyFromBytes(token[:])
+		if tokenAddr.IsOnCurve() && !tokenAddr.IsZero() {
```

Comment:
Based on the error message below, shouldn't this be 
```suggestion
		if tokenAddr.IsOnCurve() || tokenAddr.IsZero() {
```

Reply:
it should be on curve and not zero, the err msg is bad

Reply:
Just to leave a note here too for visibility. A valid token address is on the curve and is non-zero. Updated the error case accordingly.